clients/mssql/staging.go

```diff
@@ -0,0 +1,80 @@
+package mssql
+
+import (
+	"fmt"
+
+	"github.com/artie-labs/transfer/lib/config"
+
+	mssql "github.com/microsoft/go-mssqldb"
+
+	"github.com/artie-labs/transfer/lib/config/constants"
+	"github.com/artie-labs/transfer/lib/destination/ddl"
+	"github.com/artie-labs/transfer/lib/destination/types"
+	"github.com/artie-labs/transfer/lib/optimization"
+)
+
+func (s *Store) PrepareTemporaryTable(tableData *optimization.TableData, tableConfig *types.DwhTableConfig, tempTableName string, _ types.AdditionalSettings) error {
+	if tableData.Mode() != config.History {
+		tempAlterTableArgs := ddl.AlterTableArgs{
+			Dwh:               s,
+			Tc:                tableConfig,
+			FqTableName:       tempTableName,
+			CreateTable:       true,
+			TemporaryTable:    true,
+			ColumnOp:          constants.Add,
+			UppercaseEscNames: &s.config.SharedDestinationConfig.UppercaseEscapedNames,
+			Mode:              tableData.Mode(),
+		}
+
+		if err := ddl.AlterTable(tempAlterTableArgs, tableData.ReadOnlyInMemoryCols().GetColumns()...); err != nil {
+			return fmt.Errorf("failed to create temp table: %w", err)
+		}
+	}
+
+	tx, err := s.Begin()
+	if err != nil {
+		return fmt.Errorf("failed to begin transaction: %w", err)
+	}
+
+	defer func() {
+		if err != nil {
+			tx.Rollback()
+		}
+	}()
+
+	columns := tableData.ReadOnlyInMemoryCols().GetColumnsToUpdate(s.config.SharedDestinationConfig.UppercaseEscapedNames, nil)
+	stmt, err := tx.Prepare(mssql.CopyIn(tempTableName, mssql.BulkOptions{}, columns...))
+	if err != nil {
+		return fmt.Errorf("failed to prepare bulk insert: %w", err)
+	}
+
+	defer stmt.Close()
+
+	additionalDateFmts := s.config.SharedTransferConfig.TypingSettings.AdditionalDateFormats
+	for _, value := range tableData.Rows() {
+		var row []any
+		for _, col := range columns {
+			colKind, _ := tableData.ReadOnlyInMemoryCols().GetColumn(col)
+			castedValue, castErr := parseValue(value[col], colKind, additionalDateFmts)
+			if castErr != nil {
```

Comment:
If a `castErr` happens we won't roll back since it's a different variable. We could either change it to err or have a boolean called something like `committed` that we set when commit succeds.

Reply:
I like `committed` as a var so it's less of a gotcha